File: packages/mda-config/typescript/src/signature.ts
/**
 * MDA v1.0.0-rc.2 §09 — DSSE PAE and trusted signature evaluation.
 */

import { canonify } from "@truestamp/canonify";
import { verify as sigstoreVerify } from "sigstore";
import type { Bundle as SerializedBundle } from "sigstore";
import { Buffer } from "node:buffer";
import { ErrorCategory, MdaConfigError } from "./errors.js";
import type { IntegrityField } from "./integrity.js";
import {
  policyContainsDidWeb,
  sigstoreSubjectsFor,
  trustsDidWebDomain,
  type TrustPolicy,
  validateTrustPolicy,
} from "./trust-policy.js";

/** §09 — one entry of top-level `signatures[]`. */
export interface SignatureEntry {
  signer: string;
  "key-id": string;
  "payload-digest": string;
  algorithm: "ed25519" | "ecdsa-p256" | "rsa-pss-sha256";
  signature: string;
  "rekor-log-id"?: string;
  "rekor-log-index"?: number;
  "payload-type"?: string;
}

/** §09 — default DSSE payload-type when omitted. */
export const DEFAULT_PAYLOAD_TYPE = "application/vnd.mda.integrity+json";

const VENDOR_JSON_PAYLOAD_TYPE =
  /^application\/vnd\.[a-z0-9][a-z0-9-]*(?:\.[a-z0-9][a-z0-9-]*)+\+json$/u;

/** Rekor dsse-v0.0.1 entry subset needed by verifier hooks. */
export interface RekorEntry {
  kind: "dsse-v0.0.1" | string;
  logId?: string;
  logIndex?: number;
  inclusionVerified?: boolean;
  certificatePem?: string;
  dsseEnvelope: {
    payloadType: string;
    payload: string;
    signatures: { sig: string; keyid?: string }[];
  };
  rawBundle?: SerializedBundle;
}

/** Rekor lookup hook. It must fetch from the supplied policy Rekor URL. */
export interface RekorClient {
  url?: string;
  rekorUrl?: string;
  fetchEntry(rekorUrl: string, logId: string, logIndex: number): Promise<RekorEntry | null>;
}

/** Sigstore verifier hook for Fulcio chain, inclusion proof, and signature crypto. */
export interface SigstoreVerifier {
  verify(
    entry: RekorEntry,
    signature: SignatureEntry,
    paeBytes: Uint8Array,
  ): Promise<SigstoreVerificationResult>;
}

/** Verified identity returned by the Sigstore verifier hook. */
export interface SigstoreVerificationResult {
  certificatePem?: string;
  issuer?: string;
  subject?: string;
  identity?: {
    issuer?: string;
    subject?: string;
    subjectAlternativeName?: string;
  };
}

/** Input passed to a did:web cryptographic verifier hook. */
export interface DidWebVerificationInput {
  domain: string;
  keyId: string;
  algorithm: SignatureEntry["algorithm"];
  signature: string;
  payloadType: string;
  payloadBytes: Uint8Array;
  paeBytes: Uint8Array;
}

/** did:web verifier hook. Domain matching alone is never sufficient. */
export interface DidWebVerifier {
  verify(input: DidWebVerificationInput): Promise<boolean | { trusted?: boolean }>;
}

export interface SignatureVerificationOptions {
  rekorClient?: RekorClient;
  sigstoreVerifier?: SigstoreVerifier;
  didWebVerifier?: DidWebVerifier;
}

async function verifySigstoreBundle(
  bundle: SerializedBundle,
  payload: Buffer,
): Promise<SigstoreVerificationResult> {
  const signer = await sigstoreVerify(bundle, payload);
  return {
    certificatePem: extractCertFromBundle(bundle),
    identity: {
      issuer: signer.identity?.extensions?.issuer,
      subjectAlternativeName: signer.identity?.subjectAlternativeName,
    },
  };
}

/** §09 — construct DSSE PAE bytes. */
export function constructDssePae(
  payloadType: string,
  payloadBytes: Uint8Array,
): Uint8Array {
  const head = `DSSEv1 ${payloadType.length} ${payloadType} ${payloadBytes.length} `;
  const headBytes = new TextEncoder().encode(head);
  const out = new Uint8Array(headBytes.length + payloadBytes.length);
  out.set(headBytes, 0);
  out.set(payloadBytes, headBytes.length);
  return out;
}

/** §09 — JCS-canonicalize the integrity object for PAE payload bytes. */
export function paePayloadBytes(integrity: IntegrityField): Uint8Array {
  return new TextEncoder().encode(canonify(integrity));
}

/** §13 — evaluate signatures against an RC2 trust policy threshold. */
export async function verifySignatures(
  signatures: SignatureEntry[],
  integrity: IntegrityField,
  policyInput: unknown,
  options: SignatureVerificationOptions,
): Promise<void> {
  const policy = validateTrustPolicy(policyInput);
  assertVerifierHooks(policy, options);
  if (signatures.length === 0) {
    throw new MdaConfigError(
      ErrorCategory.MissingRequiredSignature,
      "trusted-runtime requires a non-empty signatures[] field",
    );
  }

  const payloadBytes = paePayloadBytes(integrity);
  const trusted = new Set<string>();
  const candidateErrors: MdaConfigError[] = [];

  for (const sig of signatures) {
    validateSignatureShape(sig);
    assertPayloadDigest(sig, integrity);
    try {
      const identity = await verifyCandidate(sig, integrity, payloadBytes, policy, options);
      if (identity) trusted.add(identity);
    } catch (cause) {
      candidateErrors.push(asMdaError(cause, ErrorCategory.SignatureVerificationFailure));
    }
  }

  const required = policy.minSignatures ?? 1;
  if (trusted.size >= required) return;
  if (trusted.size > 0) {
    throw new MdaConfigError(
      ErrorCategory.InsufficientTrustedSignatures,
      "trusted signatures did not satisfy minSignatures",
      { trusted: trusted.size, required },
    );
  }
  throw mostSpecificCandidateError(candidateErrors);
}

const defaultSigstoreVerifier: SigstoreVerifier = {
  async verify(entry, _sig, paeBytes) {
    if (!entry.rawBundle) {
      throw new Error("officialSigstoreVerifier requires RekorEntry.rawBundle");
    }
    return await verifySigstoreBundle(entry.rawBundle, Buffer.from(paeBytes));
  },
};

/** Default verifier wrapping the official sigstore npm package. */
export function officialSigstoreVerifier(): SigstoreVerifier {
  return defaultSigstoreVerifier;
}

function assertVerifierHooks(
  policy: TrustPolicy,
  options: SignatureVerificationOptions,
): void {
  if (policy.trustedSigners.some((signer) => signer.type === "sigstore-oidc")) {
    if (!options.rekorClient || !options.sigstoreVerifier) {
      throw new MdaConfigError(
        ErrorCategory.TrustPolicyViolation,
        "Sigstore trusted-runtime requires rekorClient and sigstoreVerifier hooks",
      );
    }
    const clientUrl = options.rekorClient.url ?? options.rekorClient.rekorUrl;
    if (clientUrl && clientUrl !== policy.rekor?.url) {
      throw new MdaConfigError(
        ErrorCategory.TrustPolicyViolation,
        "Rekor client URL does not match trustPolicy.rekor.url",
        { policyUrl: policy.rekor?.url, clientUrl },
      );
    }
  }
  if (policyContainsDidWeb(policy) && !options.didWebVerifier) {
    throw new MdaConfigError(
      ErrorCategory.TrustPolicyViolation,
      "did:web trusted-runtime requires a didWebVerifier hook",
    );
  }
}

async function verifyCandidate(
  sig: SignatureEntry,
  integrity: IntegrityField,
  payloadBytes: Uint8Array,
  policy: TrustPolicy,
  options: SignatureVerificationOptions,
): Promise<string | undefined> {
  if (sig.signer.startsWith("sigstore-oidc:")) {
    return await verifySigstoreCandidate(sig, payloadBytes, policy, options);
  }
  if (sig.signer.startsWith("did-web:")) {
    return await verifyDidWebCandidate(sig, payloadBytes, policy, options);
  }
  throw new MdaConfigError(
    ErrorCategory.UnknownSignerMethod,
    "unknown signer method",
    { signer: sig.signer, digest: integrity.digest },
  );
}

async function verifySigstoreCandidate(
  sig: SignatureEntry,
  payloadBytes: Uint8Array,
  policy: TrustPolicy,
  options: SignatureVerificationOptions,
): Promise<string | undefined> {
  const issuer = parseSigstoreSigner(sig.signer);
  if (sigstoreSubjectsFor(policy, issuer).size === 0) return undefined;
  const payloadType = declaredPayloadType(sig);
  const paeBytes = constructDssePae(payloadType, payloadBytes);
  const entry = await fetchRekorEntry(sig, policy, options);
  validateRekorDsseEnvelope(entry, sig, payloadType, payloadBytes);
  const verified = await options.sigstoreVerifier!.verify(entry, sig, paeBytes);
  const verifiedIssuer = verified.identity?.issuer ?? verified.issuer;
  const verifiedSubject =
    verified.identity?.subject ??
    verified.identity?.subjectAlternativeName ??
    verified.subject;
  if (!verifiedIssuer || !verifiedSubject) {
    throw new MdaConfigError(
      ErrorCategory.SignatureVerificationFailure,
      "Sigstore verifier did not return verified issuer and subject",
      { signer: sig.signer },
    );
  }
  if (verifiedIssuer !== issuer) return undefined;
  if (!sigstoreSubjectsFor(policy, verifiedIssuer).has(verifiedSubject)) return undefined;
  return `sigstore-oidc\0${verifiedIssuer}\0${verifiedSubject}`;
}

async function verifyDidWebCandidate(
  sig: SignatureEntry,
  payloadBytes: Uint8Array,
  policy: TrustPolicy,
  options: SignatureVerificationOptions,
): Promise<string | undefined> {
  const domain = parseDidWebSigner(sig.signer);
  if (!trustsDidWebDomain(policy, domain)) return undefined;
  const payloadType = declaredPayloadType(sig);
  const paeBytes = constructDssePae(payloadType, payloadBytes);
  const result = await options.didWebVerifier!.verify({
    domain,
    keyId: sig["key-id"],
    algorithm: sig.algorithm,
    signature: sig.signature,
    payloadType,
    payloadBytes,
    paeBytes,
  });
  const trusted = typeof result === "boolean" ? result : result.trusted === true;
  if (!trusted) return undefined;
  return `did-web\0${domain}`;
}

async function fetchRekorEntry(
  sig: SignatureEntry,
  policy: TrustPolicy,
  options: SignatureVerificationOptions,
): Promise<RekorEntry> {
  const rekorUrl = policy.rekor?.url;
  if (!rekorUrl) {
    throw new MdaConfigError(
      ErrorCategory.TrustPolicyViolation,
      "Sigstore trusted-runtime requires trustPolicy.rekor.url",
    );
  }
  const logId = sig["rekor-log-id"]!;
  const logIndex = sig["rekor-log-index"]!;
  const entry = await options.rekorClient!.fetchEntry(rekorUrl, logId, logIndex);
  if (!entry) {
    throw new MdaConfigError(
      ErrorCategory.RekorInclusionFailure,
      "Rekor entry not found for the supplied log coordinates",
      { logId, logIndex },
    );
  }
  if (
    entry.logId !== logId ||
    entry.logIndex !== logIndex ||
    entry.inclusionVerified !== true
  ) {
    rekorInclusionFailure("Rekor entry does not bind to signature coordinates", sig);
  }
  if (entry.kind !== "dsse-v0.0.1") {
    throw new MdaConfigError(
      ErrorCategory.RekorEntryTypeMismatch,
      "Rekor entry kind is not dsse-v0.0.1",
      { logId, logIndex, kind: entry.kind },
    );
  }
  return entry;
}

function validateSignatureShape(sig: SignatureEntry): void {
  declaredPayloadType(sig);
  if (sig.signer.startsWith("sigstore-oidc:")) {
    parseSigstoreSigner(sig.signer);
    if (!sig["rekor-log-id"] || sig["rekor-log-index"] === undefined) {
      throw new MdaConfigError(
        ErrorCategory.RekorInclusionFailure,
        "Sigstore signature requires rekor-log-id and rekor-log-index",
        { signer: sig.signer },
      );
    }
    return;
  }
  if (sig.signer.startsWith("did-web:")) {
    parseDidWebSigner(sig.signer);
    if (sig["rekor-log-id"] !== undefined || sig["rekor-log-index"] !== undefined) {
      throw new MdaConfigError(
        ErrorCategory.SchemaViolation,
        "did:web signature must not include Rekor fields",
        { signer: sig.signer },
      );
    }
    return;
  }
  throw new MdaConfigError(
    ErrorCategory.UnknownSignerMethod,
    "unknown signer method",
    { signer: sig.signer },
  );
}

function assertPayloadDigest(sig: SignatureEntry, integrity: IntegrityField): void {
  if (sig["payload-digest"] !== integrity.digest) {
    throw new MdaConfigError(
      ErrorCategory.SignatureDigestMismatch,
      "signature payload-digest does not equal integrity.digest",
      { signer: sig.signer },
    );
  }
}

function parseSigstoreSigner(signer: string): string {
  const prefix = "sigstore-oidc:";
  if (!signer.startsWith(prefix) || signer.includes("#")) {
    throw new MdaConfigError(
      ErrorCategory.UnknownSignerMethod,
      "Sigstore signer must be 'sigstore-oidc:<issuer>' with no subject suffix",
      { signer },
    );
  }
  const issuer = signer.slice(prefix.length);
  if (!issuer) {
    throw new MdaConfigError(
      ErrorCategory.UnknownSignerMethod,
      "Sigstore signer issuer is empty",
      { signer },
    );
  }
  return issuer;
}

function parseDidWebSigner(signer: string): string {
  const prefix = "did-web:";
  if (!signer.startsWith(prefix) || signer.includes("#")) {
    throw new MdaConfigError(
      ErrorCategory.UnknownSignerMethod,
      "did:web signer must be 'did-web:<domain>'",
      { signer },
    );
  }
  const domain = signer.slice(prefix.length);
  if (!domain) {
    throw new MdaConfigError(
      ErrorCategory.UnknownSignerMethod,
      "did:web signer domain is empty",
      { signer },
    );
  }
  return domain;
}

function declaredPayloadType(sig: SignatureEntry): string {
  const payloadType = sig["payload-type"] ?? DEFAULT_PAYLOAD_TYPE;
  if (payloadType.includes("+jcs+json") || !VENDOR_JSON_PAYLOAD_TYPE.test(payloadType)) {
    throw new MdaConfigError(
      ErrorCategory.SchemaViolation,
      "signature payload-type must be application/vnd.<vendor>.<doc-type>+json",
      { signer: sig.signer, payloadType },
    );
  }
  return payloadType;
}

function validateRekorDsseEnvelope(
  entry: RekorEntry,
  sig: SignatureEntry,
  payloadType: string,
  payloadBytes: Uint8Array,
): void {
  if (entry.dsseEnvelope.payloadType !== payloadType) {
    rekorInclusionFailure("Rekor DSSE envelope payloadType mismatch", sig);
  }
  if (entry.dsseEnvelope.payload !== Buffer.from(payloadBytes).toString("base64")) {
    rekorInclusionFailure("Rekor DSSE envelope payload mismatch", sig);
  }
  if (
    !entry.dsseEnvelope.signatures.some(
      (candidate) => candidate.sig === sig.signature && candidate.keyid === sig["key-id"],
    )
  ) {
    rekorInclusionFailure("Rekor DSSE envelope does not contain signature/key-id", sig);
  }
}

function rekorInclusionFailure(message: string, sig: SignatureEntry): never {
  throw new MdaConfigError(
    ErrorCategory.RekorInclusionFailure,
    message,
    { signer: sig.signer },
  );
}

function mostSpecificCandidateError(errors: MdaConfigError[]): MdaConfigError {
  return errors.find((err) => err.category === ErrorCategory.RekorEntryTypeMismatch) ??
    errors.find((err) => err.category === ErrorCategory.RekorInclusionFailure) ??
    errors.find((err) => err.category === ErrorCategory.FulcioChainFailure) ??
    errors.find((err) => err.category === ErrorCategory.SignatureVerificationFailure) ??
    new MdaConfigError(
      ErrorCategory.NoTrustedSignature,
      "no cryptographically verified signature matched the trust policy",
    );
}

function asMdaError(cause: unknown, fallback: ErrorCategory): MdaConfigError {
  if (cause instanceof MdaConfigError) return cause;
  return new MdaConfigError(
    fallback,
    "signature candidate verification failed",
    { cause: cause instanceof Error ? cause.message : String(cause) },
  );
}

function extractCertFromBundle(bundle: SerializedBundle): string {
  const chain = bundle.verificationMaterial?.x509CertificateChain?.certificates;
  if (chain?.[0]?.rawBytes) return derToPem(Buffer.from(chain[0].rawBytes, "base64"));
  const cert = (bundle.verificationMaterial as { certificate?: { rawBytes: string } })
    ?.certificate;
  if (cert?.rawBytes) return derToPem(Buffer.from(cert.rawBytes, "base64"));
  return "";
}

function derToPem(der: Buffer): string {
  const b64 = der.toString("base64");
  const lines = b64.match(/.{1,64}/gu) ?? [b64];
  return `-----BEGIN CERTIFICATE-----\n${lines.join("\n")}\n-----END CERTIFICATE-----\n`;
}

File: packages/mda-config/python/src/snoai_mda_config/signature.py
"""DSSE PAE and trusted signature evaluation."""

# pyright: reportUnusedImport=false

from __future__ import annotations

import base64
import re
from collections.abc import Sequence
from typing import NoReturn

import jcs  # type: ignore[reportMissingTypeStubs]

from .errors import ErrorCategory, MdaConfigError
from .integrity import IntegrityField
from .signature_parsing import (  # noqa: F401
    DEFAULT_PAYLOAD_TYPE,
    declared_payload_type,
    parse_did_web_signer,
    parse_sigstore_signer,
)
from .signature_types import (  # noqa: F401
    DidWebVerificationInput,
    DidWebVerifier,
    DsseEnvelope,
    DsseSignature,
    RekorClient,
    RekorEntry,
    SignatureEntry,
    SigstoreVerificationResult,
    SigstoreVerifier,
)
from .trust_policy import (
    DidWebTrustedSigner as DidWebTrustedSigner,
)
from .trust_policy import (
    SigstoreTrustedSigner as SigstoreTrustedSigner,
)
from .trust_policy import (
    TrustedSigner as TrustedSigner,
)
from .trust_policy import (
    TrustPolicy,
    policy_contains_did_web,
    sigstore_subjects_for,
    trusts_did_web_domain,
    validate_trust_policy,
)


def construct_dsse_pae(payload_type: str, payload_bytes: bytes) -> bytes:
    """MDA §09 constructs DSSE PAE bytes."""

    head = f"DSSEv1 {len(payload_type)} {payload_type} {len(payload_bytes)} ".encode()
    return head + payload_bytes


def verify_signatures(
    signatures: Sequence[SignatureEntry],
    integrity: IntegrityField,
    trust_policy: object,
    *,
    rekor_client: RekorClient | None = None,
    sigstore_verifier: SigstoreVerifier | None = None,
    did_web_verifier: DidWebVerifier | None = None,
    payload_bytes: bytes | None = None,
) -> None:
    """Evaluate signatures against a trust policy threshold."""

    policy = (
        trust_policy
        if isinstance(trust_policy, TrustPolicy)
        else validate_trust_policy(trust_policy)
    )
    _assert_verifier_hooks(policy, rekor_client, sigstore_verifier, did_web_verifier)
    if not signatures:
        raise MdaConfigError(
            ErrorCategory.MissingRequiredSignature,
            "trusted-runtime requires a non-empty signatures[] field",
        )

    payload_candidates = (
        [payload_bytes]
        if payload_bytes is not None
        else [
            _canonical_json(
                {
                    "integrity": {
                        "algorithm": integrity["algorithm"],
                        "digest": integrity["digest"],
                    }
                }
            ),
            _legacy_payload_bytes(integrity),
        ]
    )
    max_trusted = 0
    best_candidate_errors: list[MdaConfigError] = []
    best_candidate_score = -1
    saw_no_trusted_candidate = False
    for effective_payload_bytes in payload_candidates:
        trusted: set[str] = set()
        payload_errors: list[MdaConfigError] = []
        for sig in signatures:
            _validate_signature_shape(sig)
            _assert_payload_digest(sig, integrity)
            try:
                identity = _verify_candidate(
                    sig,
                    effective_payload_bytes,
                    policy,
                    rekor_client=rekor_client,
                    sigstore_verifier=sigstore_verifier,
                    did_web_verifier=did_web_verifier,
                )
                if identity is not None:
                    trusted.add(identity)
            except MdaConfigError as cause:
                payload_errors.append(cause)
            except Exception as cause:
                payload_errors.append(
                    MdaConfigError(
                        ErrorCategory.SignatureVerificationFailure,
                        "signature candidate verification failed",
                        {"cause": str(cause)},
                    )
                )

        required = policy.min_signatures
        if len(trusted) >= required:
            return
        max_trusted = max(max_trusted, len(trusted))
        if not payload_errors:
            saw_no_trusted_candidate = True
        else:
            score = _candidate_error_score(payload_errors)
            if score > best_candidate_score:
                best_candidate_score = score
                best_candidate_errors = payload_errors
    if max_trusted > 0:
        raise MdaConfigError(
            ErrorCategory.InsufficientTrustedSignatures,
            "trusted signatures did not satisfy minSignatures",
            {"trusted": max_trusted, "required": policy.min_signatures},
        )
    if saw_no_trusted_candidate:
        raise _no_trusted_signature_error()
    raise _most_specific_candidate_error(best_candidate_errors)


def _legacy_payload_bytes(integrity: IntegrityField) -> bytes:
    return _canonical_json({"algorithm": integrity["algorithm"], "digest": integrity["digest"]})


def _assert_verifier_hooks(
    policy: TrustPolicy,
    rekor_client: RekorClient | None,
    sigstore_verifier: SigstoreVerifier | None,
    did_web_verifier: DidWebVerifier | None,
) -> None:
    if any(signer.type == "sigstore-oidc" for signer in policy.trusted_signers):
        if rekor_client is None or sigstore_verifier is None:
            raise MdaConfigError(
                ErrorCategory.TrustPolicyViolation,
                "Sigstore trusted-runtime requires rekor_client and sigstore_verifier hooks",
            )
        client_url = getattr(rekor_client, "rekor_url", None)
        if client_url is not None and client_url != policy.rekor_url:
            raise MdaConfigError(
                ErrorCategory.TrustPolicyViolation,
                "Rekor client URL does not match trustPolicy.rekor.url",
                {"policyUrl": policy.rekor_url, "clientUrl": client_url},
            )
    if policy_contains_did_web(policy) and did_web_verifier is None:
        raise MdaConfigError(
            ErrorCategory.TrustPolicyViolation,
            "did:web trusted-runtime requires a did_web_verifier hook",
        )


def _verify_candidate(
    sig: SignatureEntry,
    payload_bytes: bytes,
    policy: TrustPolicy,
    *,
    rekor_client: RekorClient | None,
    sigstore_verifier: SigstoreVerifier | None,
    did_web_verifier: DidWebVerifier | None,
) -> str | None:
    if sig["signer"].startswith("sigstore-oidc:"):
        return _verify_sigstore_candidate(
            sig,
            payload_bytes,
            policy,
            rekor_client,
            sigstore_verifier,
        )
    if sig["signer"].startswith("did-web:"):
        return _verify_did_web_candidate(sig, payload_bytes, policy, did_web_verifier)
    raise MdaConfigError(
        ErrorCategory.UnknownSignerMethod,
        "unknown signer method",
        {"signer": sig["signer"]},
    )


def _verify_sigstore_candidate(
    sig: SignatureEntry,
    payload_bytes: bytes,
    policy: TrustPolicy,
    rekor_client: RekorClient | None,
    sigstore_verifier: SigstoreVerifier | None,
) -> str | None:
    issuer = parse_sigstore_signer(sig["signer"])
    payload_type = declared_payload_type(sig)
    pae_bytes = construct_dsse_pae(payload_type, payload_bytes)
    entry = _fetch_rekor_entry(sig, policy, rekor_client)
    _validate_rekor_dsse_envelope(entry, sig, payload_type, payload_bytes)
    verification = sigstore_verifier.verify(entry, sig, pae_bytes)  # type: ignore[union-attr]
    verified_issuer = verification.issuer
    verified_subject = (
        verification.subject
        or verification.subject_alternative_name
        or _extract_cert_identity(verification.certificate_pem or "")
    )
    if not verified_issuer or not verified_subject:
        raise MdaConfigError(
            ErrorCategory.SignatureVerificationFailure,
            "Sigstore verifier did not return verified issuer and subject",
            {"signer": sig["signer"]},
        )
    if verified_issuer != issuer:
        return None
    if verified_subject not in sigstore_subjects_for(policy, verified_issuer):
        return None
    return f"sigstore-oidc\0{verified_issuer}\0{verified_subject}"


def _verify_did_web_candidate(
    sig: SignatureEntry,
    payload_bytes: bytes,
    policy: TrustPolicy,
    did_web_verifier: DidWebVerifier | None,
) -> str | None:
    domain = parse_did_web_signer(sig["signer"])
    if not trusts_did_web_domain(policy, domain):
        return None
    payload_type = declared_payload_type(sig)
    ok = did_web_verifier.verify(  # type: ignore[union-attr]
        DidWebVerificationInput(
            domain=domain,
            key_id=sig["key-id"],
            algorithm=sig["algorithm"],
            signature=sig["signature"],
            payload_type=payload_type,
            payload_bytes=payload_bytes,
            pae_bytes=construct_dsse_pae(payload_type, payload_bytes),
        )
    )
    if not ok:
        return None
    return f"did-web\0{domain}"


def _fetch_rekor_entry(
    sig: SignatureEntry,
    policy: TrustPolicy,
    rekor_client: RekorClient | None,
) -> RekorEntry:
    rekor_url = policy.rekor_url
    if rekor_url is None:
        raise MdaConfigError(
            ErrorCategory.TrustPolicyViolation,
            "Sigstore trusted-runtime requires trustPolicy.rekor.url",
        )
    log_id = sig.get("rekor-log-id")
    log_index = sig.get("rekor-log-index")
    if not isinstance(log_id, str) or not isinstance(log_index, int):
        raise MdaConfigError(
            ErrorCategory.RekorInclusionFailure,
            "Sigstore signature requires Rekor coordinates",
        )
    entry = rekor_client.fetch_entry(rekor_url, log_id, log_index)  # type: ignore[union-attr]
    if entry is None:
        raise MdaConfigError(
            ErrorCategory.RekorInclusionFailure,
            "Rekor entry not found for the supplied log coordinates",
            {"logId": log_id, "logIndex": log_index},
        )
    if entry.get("log_id") is not None and entry.get("log_id") != log_id:
        _rekor_inclusion_failure("Rekor entry log ID does not match signature", sig)
    if entry.get("log_index") is not None and entry.get("log_index") != log_index:
        _rekor_inclusion_failure("Rekor entry log index does not match signature", sig)
    if entry.get("inclusion_verified") is False:
        _rekor_inclusion_failure("Rekor inclusion proof did not verify", sig)
    if entry.get("kind") != "dsse-v0.0.1":
        raise MdaConfigError(
            ErrorCategory.RekorEntryTypeMismatch,
            "Rekor entry kind is not dsse-v0.0.1",
            {"logId": log_id, "logIndex": log_index, "kind": entry.get("kind")},
        )
    return entry


def _validate_signature_shape(sig: SignatureEntry) -> None:
    declared_payload_type(sig)
    if sig["signer"].startswith("sigstore-oidc:"):
        parse_sigstore_signer(sig["signer"])
        if not sig.get("rekor-log-id") or sig.get("rekor-log-index") is None:
            raise MdaConfigError(
                ErrorCategory.RekorInclusionFailure,
                "Sigstore signature requires rekor-log-id and rekor-log-index",
                {"signer": sig["signer"]},
            )
        return
    if sig["signer"].startswith("did-web:"):
        parse_did_web_signer(sig["signer"])
        if "rekor-log-id" in sig or "rekor-log-index" in sig:
            raise MdaConfigError(
                ErrorCategory.SchemaViolation,
                "did:web signature must not include Rekor fields",
                {"signer": sig["signer"]},
            )
        return
    raise MdaConfigError(
        ErrorCategory.UnknownSignerMethod,
        "unknown signer method",
        {"signer": sig["signer"]},
    )


def _assert_payload_digest(sig: SignatureEntry, integrity: IntegrityField) -> None:
    if sig["payload-digest"] != integrity["digest"]:
        raise MdaConfigError(
            ErrorCategory.SignatureDigestMismatch,
            "signature payload-digest does not equal integrity.digest",
            {"signer": sig["signer"]},
        )


def _validate_rekor_dsse_envelope(
    entry: RekorEntry,
    sig: SignatureEntry,
    payload_type: str,
    payload_bytes: bytes,
) -> None:
    envelope = entry.get("dsse_envelope")
    if envelope is None:
        _rekor_inclusion_failure("Rekor entry is missing DSSE envelope", sig)
    if envelope["payload_type"] != payload_type:
        _rekor_inclusion_failure("Rekor DSSE envelope payloadType mismatch", sig)
    if envelope["payload"] != base64.b64encode(payload_bytes).decode("ascii"):
        _rekor_inclusion_failure("Rekor DSSE envelope payload mismatch", sig)
    if not any(
        candidate.get("sig") == sig["signature"] and candidate.get("keyid") == sig["key-id"]
        for candidate in envelope["signatures"]
    ):
        _rekor_inclusion_failure("Rekor DSSE envelope does not contain signature/key-id", sig)


def _canonical_json(value: object) -> bytes:
    raw = jcs.canonicalize(value)  # type: ignore[reportUnknownMemberType]
    return raw if isinstance(raw, bytes) else raw.encode("utf-8")


def _most_specific_candidate_error(errors: list[MdaConfigError]) -> MdaConfigError:
    for category in (
        ErrorCategory.RekorEntryTypeMismatch,
        ErrorCategory.RekorInclusionFailure,
        ErrorCategory.FulcioChainFailure,
        ErrorCategory.SignatureVerificationFailure,
    ):
        for error in errors:
            if error.category is category:
                return error
    return _no_trusted_signature_error()


def _candidate_error_score(errors: list[MdaConfigError]) -> int:
    if any(
        error.category
        in (ErrorCategory.SignatureVerificationFailure, ErrorCategory.FulcioChainFailure)
        for error in errors
    ):
        return 3
    if any(error.category is ErrorCategory.RekorEntryTypeMismatch for error in errors):
        return 2
    if any(error.category is ErrorCategory.RekorInclusionFailure for error in errors):
        return 1
    return 0


def _no_trusted_signature_error() -> MdaConfigError:
    return MdaConfigError(
        ErrorCategory.NoTrustedSignature,
        "no cryptographically verified signature matched the trust policy",
    )


def _rekor_inclusion_failure(message: str, sig: SignatureEntry) -> NoReturn:
    raise MdaConfigError(ErrorCategory.RekorInclusionFailure, message, {"signer": sig["signer"]})


def _extract_cert_identity(cert_pem: str) -> str | None:
    match = re.search(r"(?:email|uri):([^,\n]+)", cert_pem, flags=re.IGNORECASE)
    return match.group(1).strip() if match else None
